data/impl/RecordSet.ts

```diff
@@ -273,14 +274,13 @@ export class RecordSet {
     }
 
     private gatherDescendantIds(id: StoreRecordId, idSet: Set<StoreRecordId>): Set<StoreRecordId> {
-        if (!idSet.has(id)) {
-            idSet.add(id);
-            const children = this.childrenMap.get(id);
-            if (children) {
-                children.forEach(child => this.gatherDescendantIds(child.id, idSet));
+        this.childrenMap.get(id)?.forEach(child => {
+            if (!idSet.has(child.id)) {
+                // paranoia? did we encounter loops?
```

Comment:
Looks like this came in a while ago with a big refactor that introduced many of these recordset APIs - doesn't look like a response to a particular bug or use case. 

https://github.com/xh/hoist-react/pull/1082/files

(FWIW - I'd still vote to carry it forward - will cop to a certain amount of paranoia in these classes myself)